### PR TITLE
added async_closure proof of concept

### PR DESCRIPTION
**Pull request by awestlake87**
it's mostly copypasta from your #[async] attribute because I only have a vague understanding of how it works. also, just so you know, I wasn't able to figure out how to pass an Expr to your ExpandAsyncFor Folder in the case where the closure body is just an expression. I'm sure it's dead simple but I'm a total beginner when it comes to macros, let alone proc_macros.

### CODE REVIEW

File: futures-async-macro/src/lib.rs [hunk 1]
```diff
@@ -248,6 +248,99 @@ pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
...
+    block.brace_token.surround(&mut gen_body, |tokens| {
+        gen_body_inner.to_tokens(tokens);
+    });
+
+    // Give the invocation of the `gen` function the same span as the output
+    // as currently errors related to it being a result are targeted here. Not
+    // sure if more errors will highlight this function call...
+    let gen_function = quote! { ::futures::__rt::gen };
+    let body_inner = quote! {
+        #gen_function (move || #gen_body)
```
Comment (awestlake87) [inline futures-async-macro/src/lib.rs]: the `move` at line 329 should actually probably be `#capture`.

File: futures-async-macro/src/lib.rs [hunk 2]
```diff
@@ -248,6 +248,99 @@ pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
...
+        _ => panic!("async_closure! can only be applied to closures"),
+    };
+
+    let FnDecl {
+        inputs,
+        variadic,
+        ..
+    } = { *decl };
+    assert!(!variadic, "variadic functions cannot be async");
+
```
Comment (awestlake87) [inline futures-async-macro/src/lib.rs]: this next bit of code is a hack I put in to only accept closures with block bodies at the moment

**Comment by alexcrichton** [comment]
Interesting! I wonder, is a macro needed for this? Could we perhaps add a function to the prelude and do something like:

```rust
foo(|| {
    // ...
})
```

where we basically just wrap it in a generator?

**Comment by awestlake87** [comment]
At the moment, probably. The macro itself doesn't actually change much, if anything, about a function closure. But, if we were to also improve this macro to expand a generator closure into a stream future, we would likely have to change a few things. Since await! uses a void yield, and the stream generator would usually want to return an item, wouldnt we have to expand the yield type into an enum?

File: futures-async-macro/src/lib.rs (hunk 1 above)
Comment (awestlake87) [reply futures-async-macro/src/lib.rs]: actually, I take that back. it needs to be move for the closure parameters.

**Comment by awestlake87** [comment]
also, is this valid code outside of the macro?
#[async]
for i in list {
}

when I take away the async_closure! and replace it with ::futures::__rt::gen, I get this error:

error: The attribute `async` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
   --> tests/smoke.rs:197:13
    |
197 |             #[async]
    |             ^^^^^^^^
    |
    = help: add #![feature(custom_attribute)] to the crate attributes to enable

**Comment by alexcrichton** [comment]
Ah  right yeah we probably need the macro to recognize async for loops. Ok but wait now that I stop to think about this, isn't this roughly equivalent to `async_block!`?

**Comment by awestlake87** [comment]
Yeah, its pretty similar. Now that I think about it, they really are the same thing. However, does an async closure make more sense than async_block! from a semantics perspective? async_block! is essentially providing an anonymous #[async] function that closes over its environment.

**Comment by alexcrichton** [comment]
Yeah I think I'd personally lean towards keeping `async_block!` over moving to async closures as it avoids an extra layer of indentation and whatnot. And I think that an async block basically is an async closure in a sense?

**Comment by awestlake87** [comment]
Yeah, fair enough. Honestly it was great just to learn a bit about how these macros work. Im gonna go ahead and close this.

**Pull request closed without merging**